### PR TITLE
[ASTGen] Handle '#sourceLocation' directives

### PR DESCRIPTION
**Pull request by rintaro**
Use `ExportedSourceFile.sourceLocationConverter.lineTable.virtualFiles` to populate the information in `swift::SourceManger` and `swift::SourceFile` when "parsing" with ASTGen

### CODE REVIEW

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

File: lib/AST/DiagnosticBridge.cpp [hunk 1]
```diff
@@ -113,6 +113,7 @@ void *DiagnosticBridge::getSourceFileSyntax(SourceManager &sourceMgr,
     return known->second;
 
   auto bufferContents = sourceMgr.getEntireTextForBuffer(bufferID);
+  auto SF = sourceMgr.getSourceFilesForBufferID(bufferID);
```
Comment (hamishknight) [inline lib/AST/DiagnosticBridge.cpp:116]: This looks unused?

File: lib/ASTGen/Sources/ASTGen/SourceFile.swift [hunk 2]
```diff
@@ -299,3 +299,46 @@ public func findSyntaxNodeInSourceFile<Node: SyntaxProtocol>(
...
+func freeVirtualFiles(
+  cVirtualFiles: UnsafeMutablePointer<BridgedVirtualFile>?,
+  numFiles: Int
+) {
+  let buffer = UnsafeMutableBufferPointer<BridgedVirtualFile>(start: cVirtualFiles, count: numFiles)
+  for vFile in buffer {
+    freeBridgedString(bridged: vFile.Name)
+  }
+  buffer.deinitialize();
+  buffer.deallocate();
```
Comment (hamishknight) [inline lib/ASTGen/Sources/ASTGen/SourceFile.swift:343]:
```suggestion
  buffer.deinitialize()
  buffer.deallocate()
```

**Review by hamishknight — APPROVED** [review]

**Comment by rintaro** [comment]
https://github.com/swiftlang/swift-syntax/pull/2900
@swift-ci Please smoke test

**Pull request merged**